filebeat/beat/publish.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/publisher"
+)
+
+type logPublisher interface {
+	Start()
+	// TODO:
+	// Stop()
+}
+
+type syncLogPublisher struct {
+	client  publisher.Client
+	in, out chan []*input.FileEvent
+}
+
+type asyncLogPublisher struct {
+	client  publisher.Client
+	in, out chan []*input.FileEvent
+
+	// list of in-flight batches
+	active batchList
+}
+
+// eventsBatch is used to store sorted list of actively published log lines.
+// Implements `outputs.Signalerr` interface for marking batch as finished
+type eventsBatch struct {
+	next   *eventsBatch
+	flag   int32
+	events []*input.FileEvent
+}
+
+type batchList struct {
+	head, tail *eventsBatch
+}
+
+const (
+	defaultGCTimeout = 1 * time.Second
+)
+
+func newPublisher(
+	async bool,
+	in, out chan []*input.FileEvent,
+	client publisher.Client,
+) logPublisher {
+	if async {
+		return newAsyncLogPublisher(in, out, client)
+	}
+	return newSyncLogPublisher(in, out, client)
+}
+
+func newSyncLogPublisher(
+	in, out chan []*input.FileEvent,
+	client publisher.Client,
+) *syncLogPublisher {
+	return &syncLogPublisher{
+		in:     in,
+		out:    out,
+		client: client,
+	}
+}
+
+func (p *syncLogPublisher) Start() {
+	go func() {
+		logp.Info("Start sending events to output")
+
+		// Receives events from spool during flush
+		for events := range p.in {
+
+			pubEvents := make([]common.MapStr, 0, len(events))
+			for _, event := range events {
+				pubEvents = append(pubEvents, event.ToMapStr())
+			}
+
+			p.client.PublishEvents(pubEvents, publisher.Sync, publisher.Guaranteed)
+			logp.Info("Events sent: %d", len(events))
+
+			// Tell the registrar that we've successfully sent these events
+			p.out <- events
+		}
+	}()
+}
+
+func newAsyncLogPublisher(
+	in, out chan []*input.FileEvent,
+	client publisher.Client,
+) *asyncLogPublisher {
+	return &asyncLogPublisher{
+		in:     in,
+		out:    out,
+		client: client,
+	}
+}
+
+func (p *asyncLogPublisher) Start() {
+	go func() {
+		logp.Info("Start sending events to output")
+
+		// short gc timer, in case no logs are received from spooler the queued
+		// bulkEvents can still be cleaned up and forwarded to the registrar
+		ticker := time.NewTicker(defaultGCTimeout)
+
+		for {
+			select {
+			case events := <-p.in:
+				pubEvents := make([]common.MapStr, len(events))
+				for i, event := range events {
+					pubEvents[i] = event.ToMapStr()
+				}
+
+				batch := &eventsBatch{
+					flag:   0,
+					events: events,
+				}
+				p.client.PublishEvents(pubEvents,
+					publisher.Signal(batch), publisher.Guaranteed)
+
+				p.active.append(batch)
+			case <-ticker.C:
+			}
+
+			p.collect()
+		}
+	}()
+}
+
+// collect collects finished bulk-Events in order and forward processed batches
+// to registrar. Reports to registrar are guaranteed to be in same order
+// as bulk-Events have been received by the spooler
+func (p *asyncLogPublisher) collect() {
```

Comment:
by having all filebeat system tests use both modes. I think we can easily craft unit test for sync and async mode.